### PR TITLE
Dynamic CPU Load test

### PR DESCRIPTION
**Pull request by philburk**
Measure glitches when the CPU load changes.

Try to use ADPF to reduce the glitching.

### CODE REVIEW

File: apps/OboeTester/app/src/main/cpp/jni-bridge.cpp [hunk 1]
```diff
@@ -213,6 +226,15 @@ Java_com_mobileer_oboetester_OboeAudioStream_getBufferSizeInFrames(
     return result;
 }
 
+JNIEXPORT void JNICALL
+Java_com_mobileer_oboetester_OboeAudioStream_setPerformanceHintEnabled(
+        JNIEnv *env, jobject, jint streamIndex, jboolean enabled) {
+    std::shared_ptr<oboe::AudioStream> oboeStream = engine.getCurrentActivity()->getStream(streamIndex);
+    if (oboeStream != nullptr) {
+        oboeStream->setPerformanceHintEnabled(enabled);
```
Comment (robertwu1) [inline apps/OboeTester/app/src/main/cpp/jni-bridge.cpp:234]: If the user checks the checkbox before the stream is opened, would it be ignored? Should we either disable the checkbox before the stream is actually opened?

Comment (philburk) [reply apps/OboeTester/app/src/main/cpp/jni-bridge.cpp:240]: Good idea. I disabled it.

File: apps/OboeTester/app/src/main/cpp/synth/SynthTools.h [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+                                       synth_float_t *output,
+                                       int32_t numSamples) {
+        for (int i = 0; i < numSamples; i++) {
+            *output++ = *input1++ * *input2;
+        }
+    }
+
+    static double convertTimeToExponentialScaler(synth_float_t duration, synth_float_t sampleRate) {
+        // Calculate scaler so that scaler^frames = target/source
+        double numFrames = duration * sampleRate;
```
[the remark this answers is not in the document]
Comment (philburk) [reply apps/OboeTester/app/src/main/cpp/synth/SynthTools.h]: Fixed.

File: src/opensles/AudioStreamOpenSLES.cpp [hunk 3]
```diff
@@ -383,6 +383,9 @@ Result AudioStreamOpenSLES::close_l() {
     EngineOpenSLES::getInstance().close();
 
     setState(StreamState::Closed);
+
+    closePerformanceHint();
```
Comment (robertwu1) [inline src/opensles/AudioStreamOpenSLES.cpp:387]: Where is this defined for OpenSL ES? I only see a definition in AAudio

File: src/common/AdpfWrapper.cpp [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+    int result = loadAphFunctions();
+    if (result < 0) return result;
+
+    // This is a singleton.
+    APerformanceHintManager* manager = gAPH_getManagerFn();
+
+    int32_t thread32 = threadId;
+    if (sUseAlternativeHack) {
+        // FIXME hack because setprop not working!
+        targetDurationNanos = (targetDurationNanos & ~0xFF) | 0xA5;
```
Comment (robertwu1) [inline src/common/AdpfWrapper.cpp:85]: Can you add a comment of what 0xA5 represents?

File: src/aaudio/AudioStreamAAudio.cpp [hunk 5]
```diff
@@ -146,6 +146,39 @@ void AudioStreamAAudio::internalErrorCallback(
...
+}
+
+void AudioStreamAAudio::endPerformanceHintInCallback(int32_t numFrames) {
+    if (mAdpfWrapper.isOpen()) {
+        double durationScaler = static_cast<double>(mFramesPerBurst) / numFrames;
+        // Skip this callback if the numFrames is too small.
+        // This can happen when buffers wrap around, particularly when doing sample rate conversion.
+        // If we have a short callback then just scale the duration as if it was a full burst.
+        if (durationScaler < 2.0) {
+            mAdpfWrapper.onEndCallback(durationScaler);
```
Comment (robertwu1) [inline src/aaudio/AudioStreamAAudio.cpp:177]: Should we still use durationScaler when doing the scaling math here?

File: src/common/AdpfWrapper.cpp [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+    }
+}
+
+void AdpfWrapper::onBeginCallback() {
+    if (isOpen()) {
+        mBeginCallbackNanos = oboe::AudioClock::getNanoseconds(CLOCK_REALTIME);
+    }
+}
+
+void AdpfWrapper::onEndCallback(double durationScaler) {
```
Comment (robertwu1) [inline src/common/AdpfWrapper.cpp:116]: This seems to be only called when durationScaler < 2.0. Is this expected?

File: src/common/AudioStream.cpp [hunk 7]
```diff
@@ -213,4 +219,5 @@ void AudioStream::calculateDefaultDelayBeforeCloseMillis() {
          static_cast<int>(mDelayBeforeCloseMillis));
 }
 
+
```
Comment (robertwu1) [inline src/common/AudioStream.cpp:222]: Extra line

File: src/opensles/AudioStreamOpenSLES.cpp (hunk 3 above)
Comment (philburk) [reply src/opensles/AudioStreamOpenSLES.cpp:387]: I removed the call. It is not supported in OpenSL ES.

File: src/aaudio/AudioStreamAAudio.cpp (hunk 5 above)
Comment (philburk) [reply src/aaudio/AudioStreamAAudio.cpp:177]: Yes, we are calculating the scaling factor here and passing it to the ADPF wrapper which applies the scaler.
And also deciding whether we have enough data to report.

File: src/common/AudioStream.cpp (hunk 7 above)
Comment (philburk) [reply src/common/AudioStream.cpp:222]: Done

File: src/common/AdpfWrapper.cpp (hunk 6 above)
Comment (philburk) [reply src/common/AdpfWrapper.cpp:116]: It is OK to skip this if we have a short callback to reduce error. ADPF still works on sparse data.

**Review by robertwu1 — APPROVED** [review]

**Pull request merged**